jib-maven-plugin/src/main/java/com/google/cloud/tools/jib/maven/MavenLayerConfigurations.java

```diff
@@ -46,85 +43,74 @@
+
+    // Gets the resources files in the 'classes' output directory.
+    addFilesToLayer(
+        classesOutputDirectory,
+        isClassFile.negate(),
+        resourcesExtractionPath,
+        layerBuilder::addResourceFile);
 
     // Adds all the extra files.
     if (Files.exists(extraDirectory)) {
-      try (Stream<Path> extraFilesLayerDirectoryFiles = Files.list(extraDirectory)) {
-        extraFiles = extraFilesLayerDirectoryFiles.collect(Collectors.toList());
-
-      } catch (IOException ex) {
-        throw new IOException("Failed to list directory for extra files: " + extraDirectory, ex);
-      }
+      AbsoluteUnixPath extractionBase = AbsoluteUnixPath.get("/");
+      addFilesToLayer(extraDirectory, path -> true, extractionBase, layerBuilder::addExtraFile);
     }
 
-    AbsoluteUnixPath dependenciesExtractionPath =
-        appRoot.resolve(JavaEntrypointConstructor.DEFAULT_RELATIVE_DEPENDENCIES_PATH_ON_IMAGE);
-    AbsoluteUnixPath resourcesExtractionPath =
-        appRoot.resolve(JavaEntrypointConstructor.DEFAULT_RELATIVE_RESOURCES_PATH_ON_IMAGE);
-    AbsoluteUnixPath classesExtractionPath =
-        appRoot.resolve(JavaEntrypointConstructor.DEFAULT_RELATIVE_CLASSES_PATH_ON_IMAGE);
-
-    Builder layerBuilder = JavaLayerConfigurations.builder();
-    for (Path file : dependenciesFiles) {
-      layerBuilder.addDependencyFile(file, dependenciesExtractionPath.resolve(file.getFileName()));
-    }
-    for (Path file : snapshotDependenciesFiles) {
-      layerBuilder.addSnapshotDependencyFile(
-          file, dependenciesExtractionPath.resolve(file.getFileName()));
-    }
-    for (Path file : resourcesFiles) {
-      layerBuilder.addResourceFile(file, resourcesExtractionPath.resolve(file.getFileName()));
-    }
-    for (Path file : classesFiles) {
-      layerBuilder.addClassFile(file, classesExtractionPath.resolve(file.getFileName()));
-    }
-    for (Path file : extraFiles) {
-      layerBuilder.addExtraFile(file, AbsoluteUnixPath.get("/").resolve(file.getFileName()));
-    }
     return layerBuilder.build();
   }
 
+  @FunctionalInterface
+  private static interface FileToLayerAdder {
+
+    void add(Path sourcePath, AbsoluteUnixPath pathInContainer) throws IOException;
+  }
+
+  private static void addFilesToLayer(
+      Path sourceRoot,
+      Predicate<Path> pathFilter,
+      AbsoluteUnixPath basePathInContainer,
+      FileToLayerAdder addFileToLayer)
+      throws IOException {
+    Predicate<Path> isEmptyDirectory =
+        path -> Files.isDirectory(path) && path.toFile().list().length == 0;
+    // Always add empty directories. However, ignore non-empty directories because otherwise
+    // JavaLayerConfigurations will add files recursively.
+    Predicate<Path> shouldAdd =
+        path -> isEmptyDirectory.test(path) || (!Files.isDirectory(path) && pathFilter.test(path));
+
+    DirectoryWalker filteredWalker = new DirectoryWalker(sourceRoot).filter(shouldAdd);
```

Comment:
I'm thinking since we are using `Files.list` already, perhaps, we can just make this a recursive method that does `Files.list` over all the subfiles/directories and adds all the subdirectories and subfiles that pass the filter, and avoid performing a `list` on each directory while simultaneously walking the directories (resulting in essentially double listing).

Reply:
Good idea. Let me try and see how it works.

Reply:
Did the recursion version, but not sure I like it. The double listing seems more concise. In this case, I'd prefer simple code that's easy to digest and understand. I don't really think the optimization is worth.

```java
  private static void addFilesToLayer(
      Path sourceRoot,
      Predicate<Path> pathFilter,
      AbsoluteUnixPath basePathInContainer,
      FileToLayerAdder addFileToLayer)
      throws IOException {
    addFilesToLayer(sourceRoot, sourceRoot, pathFilter, basePathInContainer, addFileToLayer);
  }

  @VisibleForTesting
  static void addFilesToLayer(
      Path sourceFile,
      Path sourceRoot,
      Predicate<Path> pathFilter,
      AbsoluteUnixPath basePathInContainer,
      FileToLayerAdder addFileToLayer)
      throws IOException {
    AbsoluteUnixPath finalPath = basePathInContainer.resolve(sourceRoot.relativize(sourceFile));

    if (!Files.isDirectory(sourceFile)) {
      if (pathFilter.test(sourceFile)) {
        addFileToLayer.add(sourceFile, finalPath);
      }
      return;
    }

    List<Path> children = listDirectory(sourceFile);
    if (children.isEmpty()) { // always add empty directory
      addFileToLayer.add(sourceFile, finalPath);
    }
    for (Path child : children) {
      addFilesToLayer(child, sourceRoot, pathFilter, basePathInContainer, addFileToLayer);
    }
  }
```